plutus-tx/src/PlutusTx/Builtins/HasOpaque.hs

```diff
@@ -0,0 +1,243 @@
+
+We _also_ have the normal Haskell pair type. This is very different: you can
+pattern match on it, and you can use whatever user-defined functions you like on it.
+
+Users would really like to use the latter, and not the former. So we often want
+to _wrap_ our built-in functions with little adapters that convert between the
+"opaque builtin" "version" of a type and the "normal Haskell" "version" of a type.
+
+This is what the 'HasToOpaque' and 'HasFromOpaque' classes do. They let us write wrappers for
+builtins relatively consistently by just calling 'toOpaque' on their arguments and 'fromOpaque' on
+the result. They shouldn't really be used otherwise.
+
+Ideally, we would not have instances for types which don't have a different Haskell representation
+type, such as 'Integer'. 'Integer' in Plutus Tx user code _is_ the opaque built-in type, we don't
+expose a different one. So there's no conversion to do. However, this interacts badly with the
+instances for polymorphic built-in types, which also convert the type _inside_ them. (This is
+necessary to avoid doing multiple traversals of the type, e.g. we don't want to turn a built-in list
+into a Haskell list, and then traverse it again to conver the contents). Then we _need_ instances
+for all built-in types, so we provide a @default@ implementation for both 'toOpaque' and
+'fromOpaque' that simply returns the argument back and use it for those types that don't require any
+conversions.
+-}
+
+{- Note [HasFromOpaque/HasToOpaque instances for polymorphic builtin types]
+For various technical reasons
+(see Note [Representable built-in functions over polymorphic built-in types])
+it's not always easy to provide polymorphic constructors for built-in types, but we can usually
+provide destructors.
+
+What this means in practice is that we can write a generic 'HasFromOpaque' instance for pairs that
+makes use of polymorphic @fst@/@snd@ builtins, but we can't write a polymorphic 'ToOpaque' instance
+because we'd need a polymorphic version of the '(,)' constructor.
+
+Instead we write monomorphic instances corresponding to monomorphic constructor builtins that we add
+for specific purposes.
+-}
+
+{- Note [Fundeps versus type families in HasFromOpaque/HasToOpaque]
+We could use a type family here to get the builtin representation of a type. After all, it's
+entirely determined by the Haskell type.
+
+However, this is harder for the plugin to deal with. It's okay to have a type variable for the
+representation type that needs to be instantiated later, but it's *not* okay to have an irreducible
+type application on a type variable. So fundeps are much nicer here.
+-}
+
+-- See Note [Built-in types and their Haskell versions].
+-- See Note [HasFromOpaque/HasToOpaque instances for polymorphic builtin types].
+-- See Note [Fundeps versus type families in HasFromOpaque/HasToOpaque].
+-- | A class for converting values of transparent Haskell-defined built-in types (such as '()',
```

Comment:
So "transparent" means "pattern-matchable" here, right?

Reply:
I've added such a comment just in case.